File: src/waypalm-site/src/main/java/waypalm/site/web/extensions/CdnExtender.java
package waypalm.site.web.extensions;

import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.ui.ModelMap;
import org.springframework.web.context.request.WebRequest;
import waypalm.common.web.extensions.ModelExtension;
import waypalm.site.services.EngineSetup;

@ModelExtension("eCdn")
public class CdnExtender extends Extender<CdnExtender.Content> {
    @Autowired
    protected EngineSetup engineSetup;

    @Override
    protected CdnExtender.Content populate(WebRequest request, CdnExtender.Content extension, ModelMap model) {
        extension = instantiateIfNull(extension, CdnExtender.Content.class);
        extension.debugMode = engineSetup.isDebugMode();
        extension.server = engineSetup.getSiteUrl();
        extension.serverStart = engineSetup.getStartTime();
        extension.contextPath = engineSetup.getServletContext();
        extension.locale = request.getLocale().getLanguage();
        return extension;
    }

    public static class Content {
        private int controlId;

        private boolean debugMode;
        private String server;
        private long serverStart;
        private String contextPath;
        private String locale;
        private String appZone;

        public long getServerStart() {
            return debugMode ? System.currentTimeMillis() : serverStart;
        }

        public String getServer() {
            return server;
        }

        public String getContextPath() {
            return contextPath;
        }

        public String getAppZone() {
            return appZone;
        }

        public void setAppZone(String appZone) {
            this.appZone = appZone;
        }

        public String getLocale() {
            return locale;
        }

        public boolean isDebugMode() {
            return debugMode;
        }

        public String getNextCid() {
            ++controlId;
            return getCid();
        }

        public String getCid() {
            return "c" + controlId;
        }
    }
}

File: src/waypalm-site/src/main/java/waypalm/site/web/extensions/CdnExtender.kt
package waypalm.site.web.extensions

import waypalm.common.web.extensions.ModelExtension
import javax.inject.Inject
import waypalm.site.services.EngineSetup
import org.springframework.web.context.request.WebRequest
import org.springframework.ui.ModelMap

ModelExtension("eCdn")
public class CdnExtender [Inject](
        var engineSetup: EngineSetup
)
: Extender<CdnExtender.Content>()
{
    protected override fun populate(request: WebRequest, extension: CdnExtender.Content?, model: ModelMap): CdnExtender.Content {
        var res = instantiateIfNull(extension, javaClass<Content>());
        res.debugMode = engineSetup.debugMode;
        res.server = engineSetup.siteUrl
        res.serverStart = engineSetup.startTime
        res.contextPath = engineSetup.servletContext
        res.locale = request.getLocale()!!.getLanguage();
        return res;
    }

    public class Content {
        var controlId: Int = 0
        var debugMode: Boolean = false
        var server: String = ""
        var serverStart: Long = 0
            get() = if(debugMode) System.currentTimeMillis() else $serverStart

        var contextPath: String = ""
        var locale: String = ""
        var appZone: String = ""

        fun getNextCid(): String {
            ++controlId
            return getCid()
        }

        fun getCid(): String {
            return "c${controlId}"
        }
    }
}